### PR TITLE
Sign up

### PR DESCRIPTION
**Pull request by MimoMoreno**
## Summary
* The sign up was implemented.
* To create a new user make a POST request to the /users endpoint.
* The test for this feature was implemented but it is not working due to a database error.

## Trello Card
https://trello.com/c/ZLamAhJ6

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  if (user.password.length < 8) {
+    return next(errors.validationError('The password is too short. It must have more than 7 characters.'));
+  }
+
+  bcrypt.hash(user.password, saltRounds).then((hash) => {
+    // creates user
+    user.password = hash;
+    userService.create(user).then(() => {
+      res.status(200);
+      res.send('User created successfully.');
```
Comment (mishuagopian) [inline app/controllers/users.js]: No need to answer with a message like this. Just close the responde with `res.end();`

File: app/controllers/users.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+const bcrypt = require('bcrypt'),
+  userService = require('../services/users.js'),
+  errors = require('../errors');
+
+exports.create = (req, res, next) => {
+  const saltRounds = 10;
+
```
Comment (mishuagopian) [inline app/controllers/users.js]: remove empty line

File: app/controllers/users.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  const user = {
+    firstName: req.body.firstName,
+    lastName: req.body.lastName,
+    email: req.body.email,
+    password: req.body.password
+  };
+
+  // checks email
```
Comment (mishuagopian) [inline app/controllers/users.js]: remove comments

File: app/controllers/users.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  bcrypt.hash(user.password, saltRounds).then((hash) => {
+    // creates user
+    user.password = hash;
+    userService.create(user).then(() => {
+      res.status(200);
+      res.send('User created successfully.');
+    }).catch((err) => {
+      return next(errors.databaseError(err.errors));
```
Comment (mishuagopian) [inline app/controllers/users.js]: This error (dbError) should be caught inside the service, as the controller does not know about db's existance, turning this catch into `.catch(next);` or `.catch((err) => next(err));`

File: app/controllers/users.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    });
+  });
```
Comment (mishuagopian) [inline app/controllers/users.js]: What if bcrypt fails?

File: app/services/users.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+'use strict'
+const orm = require('./../orm.js');
+
+exports.create = (user) => {
+  return orm.models.user.create({
+    firstName: user.firstName,
+    lastName: user.lastName,
+    email: user.email,
+    password: user.password
+  });
```
Comment (mishuagopian) [inline app/services/users.js]: Any db error?

File: test/app.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  db.sync({ force: true }).then(() => {
+    exports.models = db.models;
+    done();
+  }).catch((err) => {
+    console.log('hello');
+    console.log(err);
+  });
+});
+
+// beforeEach('populate tables', (done) => {
```
Comment (mishuagopian) [inline test/app.js]: This shouldn't be commented

File: app/models/scripts/dataCreation.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+      lastName: 'lastName1',
+      email: 'email1@wolox.com.ar',
+      password: hash
+    }));
+    users.push(db.models.user.create({
+      firstName: 'firstName2',
+      lastName: 'lastName2',
+      email: 'email2@wolox.com.ar',
+      password: hash
+    }));
```
Comment (mishuagopian) [inline app/models/scripts/dataCreation.js]: You should return `Promise.all(users);`

File: test/users.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
+const chai = require('chai'),
+  should = chai.should(),
+  server = require('./../app');
+
+describe('users', () => {
+  describe('/users POST - users creation', () => {
```
Comment (mishuagopian) [inline test/users.js]: Just `'/users POST'`

File: test/users.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    it('should fail because of invalid email', (done) => {
+      chai.request(server)
+        .post('/users')
+        .send({ firstName: 'Mishu', lastName: 'Agopian', email: 'mishuu@wolox.com', password: 'password1234' })
+        .catch((err) => {
+          err.should.have.status(400);
+          err.response.should.be.json;
+          err.response.body.should.have.property('error');
+        }).then(() => done());
+    });
```
Comment (mishuagopian) [inline test/users.js]: I know tests are failing right now, but we should have more test cases. At least:
- invalid/in use username.
- in use email.

File: app/models/scripts/dataCreation.js (hunk 8 above)
Comment (MimoMoreno) [inline app/models/scripts/dataCreation.js]: `users` cannot be accessed from the `return` statement. The `users` object is inside the `data`.

Comment (mishuagopian) [inline app/models/scripts/dataCreation.js]: What? I think you must return users or you may start testing without the users in the db

**Comment by mishuagopian** [comment]
@MimoMoreno please fix this PR !

**Pull request merged**